core/node/base_token_adjuster/src/base_token_l1_behaviour.rs

```diff
@@ -0,0 +1,331 @@
+use std::{
+    cmp::max,
+    ops::{Div, Mul},
+    sync::Arc,
+    time::Instant,
+};
+
+use anyhow::Context;
+use bigdecimal::{num_bigint::ToBigInt, BigDecimal, Zero};
+use zksync_config::BaseTokenAdjusterConfig;
+use zksync_eth_client::{BoundEthInterface, CallFunctionArgs, Options};
+use zksync_node_fee_model::l1_gas_price::TxParamsProvider;
+use zksync_types::{
+    base_token_ratio::BaseTokenAPIRatio,
+    ethabi::{Contract, Token},
+    web3::{contract::Tokenize, BlockNumber},
+    Address, U256,
+};
+
+use crate::metrics::{OperationResult, OperationResultLabels, METRICS};
+
+#[derive(Debug, Clone)]
+pub struct UpdateOnL1Params {
+    pub eth_client: Box<dyn BoundEthInterface>,
+    pub gas_adjuster: Arc<dyn TxParamsProvider>,
+    pub token_multiplier_setter_account_address: Address,
+    pub chain_admin_contract: Contract,
+    pub getters_facet_contract: Contract,
+    pub diamond_proxy_contract_address: Address,
+    pub chain_admin_contract_address: Option<Address>,
+    pub config: BaseTokenAdjusterConfig,
+}
+
+#[derive(Debug, Clone)]
+pub enum BaseTokenL1Behaviour {
+    UpdateOnL1 {
+        params: UpdateOnL1Params,
+        last_persisted_l1_ratio: Option<BigDecimal>,
```

Comment:
does it need to be an `Option`? It's only `None` on startup, so we do initialization implicitely inside normal logic - and have a slightly more complicated state space (needing `Option` now). Could we request the persisted l1 ration on startup (in a `new` method) instead?

Well you'd probably still want to have several attempts of requesting it from l1, which may be complicated - but maybe not, especially if retry logic is abstracted away.

Ok, so overall not critical - feel free to leave as is

Reply:
Good point - let me make a ticket for that.